---
title: "Record Marching of identical individuals"
output: rmarkdown::html_vignette
author: "Jaehee Kim"
date: "05/28/2021"
vignette: >
  %\VignetteIndexEntry{identical}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

---

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
devtools::load_all()
library(knitr)
```


```{r setup}
library(RecordMatching)
```

## Overview
This vignette demonstrates record-matching between SNP profiles to STR profiles when two samples are from the identical individual using linkage disequilibrium between samples. 

In this example, we will use the following 13 CODIS loci.
```{r}
marker.f <- system.file("extdata", "marker_positions.txt",
                        package="RecordMatching", mustWork=TRUE)
marker.info <- read.table(marker.f, header=TRUE, as.is=TRUE)
```

```{r echo=FALSE}
kable(marker.info)
```

One can include specific loci only by subsetting the `marker.info` as shown below. For a quick testing of the pipeline, start with one locus. In general, including more loci results in the better record-matching accuracy. 

```{r}
loci.include <- c(1)
marker.info <- marker.info[loci.include,]
```

```{r echo=FALSE}
cat("Loci chosen:", marker.info$name)
```



## Setup
### External software requirements
In order to run the pipeline, there are two external software requires.

* [BEAGLE 4.1](https://faculty.washington.edu/browning/beagle/b4_1.html) for phasing and imputation.
  Please note that `BEAGLE` requires Java version 8. See [BEAGLE 4.1 manual](https://faculty.washington.edu/browning/beagle/beagle_4.1_21Jan17.pdf) for details.
* [VCFtools](https://github.com/vcftools/vcftools) for extracting necessary information from vcf files.


### Set environment variables
We first need to set the base directory where the output files of the record matching pipeline will be saved. We also set the full path to the `BEAGLE` jar file and `vcftools` executable. The following environment variables need to be set.

* `base.dir`: This is a directory where all the files generated by the pipeline will be saved.
* `bgl.jar`: Path to `BEAGLE` jar file. Format is `(full path to dir)/beagle.(version).jar`. `(version)` is the Beagle version code (eg. “01Oct15.6a3”).
* `vcf.exe`: Path to `vcftools` executable. Format is `(full path to dir)/vcftools`.

The above variables will be used in all functions in the record-matching pipeline. To run this vignette, change the paths below to  


```{r}
setup(base.dir='/Users/jaehee/Desktop/RM_prog/twin_test',
      bgl.jar='/Users/jaehee/Desktop/RM_prog/beagle.27Jan18.7e1.jar',
      vcf.exe='/usr/local/bin/vcftools')
```


### Input files requirements

In addition, 







## Phase SNP-STR reference panel
**The reference panel must be phased and have no missing values**.


Phase reference panel and extract allele frequencies at STR loci.  
If the reference panel is unphased, it must be phased first using 

```{r}
for (m in marker.info$name) {
    ref.f <- system.file("extdata", "twin_test", "reference",
                         paste0("ref_", m, ".vcf"),
                         package="RecordMatching", mustWork=TRUE)

    # if reference panel is unphased, phase them. 
    phase.ref(ref.f)

    # otherwise, go straight to allele frequency
    ref.al.freq(ref.f, m)
}
```



## Impute reference panel and compute imputed genotype probabilities


```{r, eval=FALSE}
for (i in 1:dim(marker.info)[1]) {
    m <- marker.info$name[i]
    chr <- marker.info$chr[i]
    snp.f <- system.file("extdata", "twin_test", "SNP",
                         paste("snp_", m, ".vcf", sep=''),
                         package="RecordMatching", mustWork=TRUE)
    ref.f <- system.file("extdata", "twin_test", "reference",
                         paste("ref_", m, ".vcf", sep=''),
                         package="RecordMatching", mustWork=TRUE)
    map.f <- system.file("extdata", "twin_test", "plink.GRCh36.map",
                         paste("plink.chr", chr, '.GRCh36.map', sep=''),
                         package="RecordMatching", mustWork=TRUE)

    impute.str(snp.f=snp.f, ref.f=ref.f, map.f=map.f, marker=m)
}
```



## Compute match score matrix 
Iterate over markers and add them to get the final match score matrix using 13 CODIS loci. 

```{r}
cott.vec <- c(0,0,1) # same individual
# cott.vec <- c(0.25, 0.5, 0.25) # sib pairs
# cott.vec <- c(0,1,0) # parent-offspring

llr.single.irt <- list()

for (i in 1:dim(marker.info)[1]) {
    m <- marker.info$name[i]
    print(paste("Processing marker", m, ",", i, "out of", length(marker.info$name)))
    str.f <- system.file("extdata", "twin_test", "STR",
                         paste("str_", m, ".GT.FORMAT", sep=''),
                         package="RecordMatching", mustWork=TRUE)
    llr.single.irt[[i]] <- comp.match.mat(cott.vec=cott.vec, marker=m, str.f=str.f)
    names(llr.single.irt)[i] <- m
}

mat <- Reduce('+', llr.single.irt)
```


The match score matrix using 
```{r mat-plot, dev='png', echo=FALSE, out.width="50%", fig.align='center'}
par(mar=c(1,1,0,0))
image(mat, axes=FALSE, asp=1, frame.plot=FALSE)
mtext('STR Profile', side=1)
mtext('SNP Profile', side=2)
```


Compute match accuracies.
```{r}
match.acc <- comp.match.acc(mat)
print(match.acc)
```





